Add polarizer efficiency algorithm

jclarkeSTFC opened the pull request:
New algorithm to calculate the supermirror polarizer efficiency using a group workspace with 4 runs for the 4 spin combinations, plus the efficiency of the analyser cell, calculated previously. This is part of the polarized SANS epic.

Output can either be to a workspace, a file, or both.

Fixes #36139

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
- Added algorithm
- Added tests
- Added doc page

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
Calculates how the efficiency of a polarizer varies with wavelength. The ordering of the workspaces in ``InputWorkspace`` is taken from the ``SpinStates`` parameter, and the analyser efficiency, e_cell, is given by ``AnalyserEfficiency``.

The polarization of the polarizer, P_SM, is given by

![image](https://github.com/mantidproject/mantid/assets/139879523/695c219e-f5d6-41e4-a0bc-1a5a6d4f32c2)

Since the efficiency, e_SM , is given by (1 + P_SM)/2, we have that

![image](https://github.com/mantidproject/mantid/assets/139879523/2ef7f8f2-b461-48f2-bfa3-26d564101f28)

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->
This will generate a simple example:
``` Python
wsPara = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1-0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
wsPara1 = CloneWorkspace(wsPara)
wsAnti = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=0.5*exp(-0.0733*12*x*(1+0.9))',XUnit='Wavelength', xMin='1',XMax='8', BinWidth='1', NumBanks='1', BankPixelWidth='1')
wsAnti1 = CloneWorkspace(wsAnti)

grp = GroupWorkspaces([wsPara,wsAnti,wsPara1,wsAnti1])
eCell = CreateSampleWorkspace('Histogram', Function='User Defined', UserDefinedFunction='name=UserFunction,Formula=(1 + tanh(0.0733 * 12 * x * 0.2))/2',XUnit='Wavelength', xMin='1',XMax='16', BinWidth='1', NumBanks='1', BankPixelWidth='1')

psm = PolarizerEfficiency(grp, eCell)
print("Polarizer efficiency at a wavelength of " + str(mtd['psm'].dataX(0)[3]) + " Å is " + str(mtd['psm'].dataY(0)[3]))
```

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Add algorithm, tests, and docs

Calculate polarizer efficiency using a group workspace with
4 runs for the 4 spin combinations, plus the efficiency of the
analyser cell, calculated previously.

jclarkeSTFC committed:
Fix doc test

jclarkeSTFC committed:
Update to use new formula for efficiency

jclarkeSTFC committed:
Pre-emptive attack on Coverity

cailafinn reviewed (changes requested):
### Code Review:
- Have a few code notes, but it's looking good. 

### Functional testing:
- Working well from both the script and the algorithm dialog. Tried a few different bad inputs and it handled them well. 
- The 1st column of the output data from the script you provided seems to provide an efficiency value of `-1.108...`. I was under the impression efficiencies should always be between 0 and 1, so I just want to check if that's expected or if I've not got confused about something. 
 
### Docs:
- Look great! Thanks for including the maths for both the polarisation and the efficiency.

jclarkeSTFC committed:
Tidy up code

Missing const in a few places, deleted unneeded methods and used the
operator overloads instead.

jclarkeSTFC committed:
Update formula for polarizer efficiency

Includes updating expected test results

jclarkeSTFC commented:
> ### Code Review:
> 
>     * Have a few code notes, but it's looking good.
> 
> 
> ### Functional testing:
> 
>     * Working well from both the script and the algorithm dialog. Tried a few different bad inputs and it handled them well.
> 
>     * The 1st column of the output data from the script you provided seems to provide an efficiency value of `-1.108...`. I was under the impression efficiencies should always be between 0 and 1, so I just want to check if that's expected or if I've not got confused about something.
> 
> 
> ### Docs:
> 
>     * Look great! Thanks for including the maths for both the polarisation and the efficiency.

Thanks, I think I've changed all the things you spotted, and I've updated to the new equation.

jclarkeSTFC committed:
Fix cppcheck

jclarkeSTFC committed:
Update equations in docs

jclarkeSTFC committed:
Add option to save to file

You can now either have an output workspace, or save to a file, or both.

cailafinn commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,185 @@
+ */
+std::map<std::string, std::string> PolarizerEfficiency::validateInputs() {
+  std::map<std::string, std::string> errorList;
+  const std::string inputWorkspaceName = getProperty(PropertyNames::INPUT_WORKSPACE);
+  if (!AnalysisDataService::Instance().doesExist(inputWorkspaceName)) {
+    errorList[PropertyNames::INPUT_WORKSPACE] = "The workspace " + inputWorkspaceName + " does not exist in the ADS.";
+    return errorList;
+  }
```
I don't think the workspace needs to be in the ADS for the algorithm to work. If it's going to be used as a child algorithm (like we plan to do with the workflow algorithm to cover the whole correction process) then the input workspaces are likely going to not be in the ADS. You can get the workspace directly from the input by doing something like this:
```suggestion
  const WorkspaceGroup_sptr inputWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
  if (inputWorkspace == nullptr) {
    errorList[PropertyNames::INPUT_WORKSPACE] = "The input workspace  is not a workspace group."
    return errorList;
  }
```

cailafinn commented on docs/source/algorithms/PolarizerEfficiency-v1.rst:
```diff
@@ -0,0 +1,54 @@
+
+.. math::
+    \epsilon_{SM} = \frac{1}{2} + \frac{T_{00} - T_{01}}{4(2\epsilon_{cell} - 1)(T_{00} + T_{01})}
+
```
I noticed in the code that one of the workspaces is rebinned during the calculation. Could we add a note in the docs here to mention which workspace's binning is the one that's used for the output?

cailafinn reviewed (changes requested):
Functionally working well, just a couple more things I noticed. Apologies for not catching them in the first review.

jclarkeSTFC committed:
Specify input is a workspace group

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>

jclarkeSTFC committed:
Input may not be in the ADS

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>

jclarkeSTFC committed:
Remove validator on input workspace

This validator does not work when the type is specified explicitly as
being a WorkspaceGroup, so instead we'll have to to check for wavelength
in validateInputs().

rbauststfc committed:
Add note about rebinning to documentation

rbauststfc committed:
Minor refactoring

cailafinn reviewed (changes requested):
Docs built and looking good. Code has a few more notes. Functionally it's working well for the normal use case, but there's a crash to desktop (segfault) when using an invalid workspace type for the Cell efficiency workspace.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,188 @@
+
+void PolarizerEfficiency::calculatePolarizerEfficiency() {
+  // First we extract the individual workspaces corresponding to each spin configuration from the group workspace
+  const auto &groupWorkspace =
+      AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(getProperty(PropertyNames::INPUT_WORKSPACE));
```
On this, and relevant to the other comment about simplifying the validation, have we done some testing to see if you really can pass in a workspace that's not in the ADS to a `WorkspaceGroup` property? I did what I consider to be some very incomplete testing, but it seemed to show that the property type was enforcing that the workspace group had to be in the ADS. It would be good to establish this conclusively, as it will be relevant to all the other polarised SANS algorithms.

cailafinn replied:
Yes, I can confirm this. Below shows the debugging from C++ in the tests. The `CreateSampleWorkspace` outputs are present in the ADS due to it being a non-child alg, but the group is not as the `GroupWorkspaces` alg is a child algorithm. The group only exists as a pointer in the C++ code. It's then passed to the `FlipperEfficiency` alg without entering the ADS which is able to process normally. 
![image](https://github.com/mantidproject/mantid/assets/47181718/41bc272f-d1e3-43f0-9ee1-8fc109561601)

rbauststfc replied:
OK, brilliant, thanks for checking 👍

rbauststfc commented:
Just making a note that because we'll now be targeting 6.11 for this change, we'll need to rebase and move the release note to the appropriate 6.11 folder.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp:
```diff
@@ -0,0 +1,188 @@
+  const auto &t00Ws = PolarizationCorrectionsHelpers::workspaceForSpinState(groupWorkspace, spinConfigurationInput,
+                                                                            SpinStateValidator::ZERO_ZERO);
+
+  auto effCell = convertToHistIfNecessary(
+      AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(getProperty(PropertyNames::ANALYSER_EFFICIENCY)));
```
For when we come back to this, I think the suggestion here was intended to be:

`auto effCell = convertToHistIfNecessary(getProperty(PropertyNames::ANALYSER_EFFICIENCY));`
